### PR TITLE
Check a PR has been committed using git signoff

### PR DESCRIPTION
**Pull request by NvTimLiu**
1. Pre-merge build fails without seeing a `-s` signoff from anyone committing to a PR.

    If the author of a commit adds a single commit with a `-s`, that will be sufficient for the pre-merge build to pass.

2, Only update Jenkins scripts, no source change. No unit tests needed.

3, No Github issue related to the PR.

### CODE REVIEW

**Comment by NvTimLiu** [comment]
build

File: jenkins/Jenkinsfile.premerge [hunk 1]
```diff
@@ -72,3 +78,24 @@ pipeline {
...
+boolean isSignedOff() {
+    def target_rev = sh(returnStdout: true,
+            script: "git rev-parse refs/remotes/origin/${ghprbTargetBranch}").trim()
+    def revs_arr = sh(returnStdout: true,
+            script: "git log ${target_rev}..${ghprbActualCommit} --pretty=format:%h").split()
+
+    def signed_off = false
+    for( String commit : revs_arr ) {
+        def signed_log = sh(returnStdout: true,
+                script: "git show ${commit} --shortstat | grep 'Signed-off-by'")
```
Comment (revans2) [inline jenkins/Jenkinsfile.premerge:91]: So does this mean that if I do one commit with no sign off, and then I do another commit with one that this will pass?

File: jenkins/Jenkinsfile.premerge [hunk 2]
```diff
@@ -51,6 +51,12 @@ pipeline {
             agent { label 'docker-gpu' }
             steps {
                 script {
+                    //Check if a PR has been committed using git signoff
+                    if (!isSignedOff()) {
+                        echo "Signed-off-by check FAIED"
```
Comment (jlowe) [inline jenkins/Jenkinsfile.premerge:56]:
```suggestion
                        echo "Signed-off-by check FAILED"
```

**Comment by revans2** [comment]
@NvTimLiu what are the plans for this? I'd like to see it go in, and I am happy to fix the spelling issue later if you don't have time right now to do it.

**Comment by revans2** [comment]
build

**Comment by jlowe** [comment]
@revans2 are we going with only one signoff commit required per PR?  Seems this logic would greenlight a PR with 20 commits where only 1 commit has the signoff.

**Comment by revans2** [comment]
@jlowe I am fine with that, and we can change it in the future if we need to.

**Comment by tgravescs** [comment]
"Any contribution which contains commits that are not signed off will not be accepted."


https://github.com/NVIDIA/spark-rapids/blob/branch-0.2/CONTRIBUTING.md

**Comment by revans2** [comment]
So I cannot upmerge without signing off on the merge commit?  What if the github merge commit did not have a sign-off on it? Think about the case of merging a commit from one branch to another. I understand what the document says.  We wrote that document.

**Comment by jlowe** [comment]
> What if the github merge commit did not have a sign-off on it?

Shouldn't be too long before all the commits pulled in by a merge will have signoffs, but it could be a problem in the short term.

Agree this is better than what we have today which is no checks.  Let's go with this and file a followup to either tighten the checks or update the docs.

**Comment by tgravescs** [comment]
I'm definitely fine with that as like you said its better the no checks

**Comment by sameerz** [comment]
build

**Comment by revans2** [comment]
build

**Comment by jlowe** [comment]
Appears the CI build failed because the last commit doesn't have a signoff:
```
08:05:28  + git log e129f1b43f5a1f1861e10701214ef496506fe58f..a0f557a51c2f80da8e3c42dd424a0d2ae89446cf --pretty=format:%h
08:05:28  [Pipeline] sh
08:05:29  + git show a0f557a --shortstat
08:05:29  + grep Signed-off-by
[...]
08:05:29  [Pipeline] End of Pipeline
08:05:29  Setting status of a0f557a51c2f80da8e3c42dd424a0d2ae89446cf to FAILURE with url https://ci.ngcc.nvidia.com/job/spark/job/rapids_premerge-github/315/ and message: 'Build finished. '
```

I'll dig a bit into why it didn't see the first commit in this PR which does have a signoff.

File: jenkins/Jenkinsfile.premerge (hunk 1 above)
Comment (jlowe) [inline jenkins/Jenkinsfile.premerge:91]: I think there's an issue here where if `grep` fails to find a match it will return a non-zero exit code.  That in turn will cause the pipeline command to return a non-zero exit code which will cause the `sh` command to throw an exception as described [here](https://www.jenkins.io/doc/pipeline/steps/workflow-durable-task-step/#sh-shell-script).  I think this either needs to use the `returnStatus: true` option and switch off the result code of the step or do something like this to avoid `grep` failing to match to cause a pipeline failure:
```suggestion
                script: "git show ${commit} --shortstat | grep 'Signed-off-by || echo'")
```

**Comment by NvTimLiu** [comment]
build

**Comment by NvTimLiu** [comment]
build

**Comment by NvTimLiu** [comment]
Sorry for the delayed response, I was off for personal issue last week.

Fixed the `sh` exception when `grep` not find a match of `Signed-off-by` in the commit lists of a PR.

@jlowe @revans2  Do you mean we need to make sure the "Signed-off-by" tag exists in every commit?

I've learned from Sameer @sameerz  that we can PASS the pre-merge build only if one of the PR commits(instead of all commits) have a `Signed-off-by` tag.
